### PR TITLE
Replace `tornado` dependency with built-in module `asyncio`

### LINKED ISSUE #3184: Remove upper limit for tornado dependency by replacing it with asyncio
The current jupyter-notebook (version 6.0.0) is incompatible with AiiDA 1.0.0x.

`pip install -e ".[notebook]"` yields the following errors:
```
ERROR: notebook 6.0.0 has requirement pyzmq>=17, but you'll have pyzmq 16.0.4 which is incompatible.
ERROR: notebook 6.0.0 has requirement tornado>=5.0, but you'll have tornado 4.5.3 which is incompatible.
ERROR: jupyter-console 6.0.0 has requirement prompt-toolkit<2.1.0,>=2.0.0, but you'll have prompt-toolkit 1.0.16 which is incompatible.
```
plumpy is incompatible with tornado 5 (the `stack_context` module has been removed).

### PR DESCRIPTION
**Pull request by unkcpz**
Fixes #3184 

PR #4218 is rebased here for easy to review.

### CODE REVIEW

**Comment by unkcpz** [comment]
> Also, the build on jenkins is failing. It seems the daemon is failing to start. Do you think this is a one-time fluke or there really is a problem? 

I think it is result from `circus`, it is failed because the jenkins install aiida_core with `pip install --user .[all]` which installed the old `circus`. I guess it will be fixed after new version circus released. This issue also happened in test-install and ci-test, I workaround it by install circus of my repository.  https://github.com/aiidateam/aiida-core/blob/0e1ecfbaa08f3fe0341d4083c693bf13b8884bf9/.github/workflows/ci.yml#L125

**Review by unkcpz — COMMENTED** [review]
Thanks! @sphuber Some of comments really the code I am also not very sure about the changes. Especially the `unwrap of the future in `cmd_process`, I need to think it over and get back to you.

File: aiida/cmdline/commands/cmd_process.py
Comment (unkcpz) [reply aiida/cmdline/commands/cmd_process.py:352]: You are right that this part is not very clear. And I think there is the same issue for the `aiida_core` of the tornado version. 

Let me try to describe my understanding of this part of the code:
The future returned from the `controller.kill_process` is not  depend on the communicator we used here, but on communicator which used to actually doing the `kill` action. That is the `LoopCommunicator` and we add a rpc_subscriber in that communicator of the daemon `Runner`. 
The reason that we do not need use the LoopCommunicator in `verdi process` is every time we run a cmd command we run it in a independent thread and wait for the returned value. 
And it is because we don't know what communicator we used exactly, we need to put a generic `unrap_kiwi_future` here to get the final result of the procedure scheduled in the communicator.

File: environment.yml
Comment (sphuber) [reply environment.yml]: Typically, we just update the `setup.json`. The `environment.yml` is then automatically updated by a pre-commit hook. The `requirements` files are also autogenerated, but by a Github Action workflow. The reason that this is not done by a pre-commit hook, is because it needs a complete clean and well controlled environment, so that's why it is done on the CI platform and not locally.

File: tests/engine/test_rmq.py
Comment (sphuber) [reply tests/engine/test_rmq.py:84]: ok, fair enough

File: tests/engine/test_work_chain.py
Comment (sphuber) [reply tests/engine/test_work_chain.py]: Let's remove it then

File: aiida/engine/processes/calcjobs/tasks.py
Comment (unkcpz) [reply aiida/engine/processes/calcjobs/tasks.py:377]: It is there return but by catch the situation as `gen.Return`.

File: environment.yml
Comment (unkcpz) [reply environment.yml]: Thanks! I changed every file every time before, it's really not necessary :p

File: tests/engine/test_work_chain.py
Comment (unkcpz) [reply tests/engine/test_work_chain.py]: removed.

File: aiida/cmdline/commands/cmd_process.py
Comment (muhrin) [reply aiida/cmdline/commands/cmd_process.py:352]: Haha, yes, I'm also getting confused looking through this part of the code.  I guess the principle that Sebastiaan is pointing out (and I agree with) is that if there is a `Communicator` interface that we're using then things that come out of the interface should always be at the same level of 'unwrapping', ideally such that no manual unwrapping needs to be done outside, and certainly in a way that doesn't vary on the underlying concrete `Communicator` instance type.  

I'll have a look through the code and see if I can make sure I understand what the flow is.

File: aiida/engine/daemon/runner.py
Comment (muhrin) [reply aiida/engine/daemon/runner.py:42]: This looks like a neat solution, I like the look of it.

File: aiida/engine/processes/functions.py
Comment (unkcpz) [reply aiida/engine/processes/functions.py:116]: The following docstring is there before I make the changes. This is what I mean 'blocking issues', I guess this was why a new event loop created in the ProcessFunction.

https://github.com/aiidateam/aiida-core/blob/5977dacc6160a68bba6fc4fd91c4bfa17f330520/aiida/engine/processes/functions.py#L110-L112

File: aiida/engine/daemon/runner.py
Comment (unkcpz) [reply aiida/engine/daemon/runner.py:42]: @muhrin Thanks! that is one thing I'm not sure about, now I don't have to worry.

File: tests/cmdline/commands/test_data.py
Comment (unkcpz) [reply tests/cmdline/commands/test_data.py:315]: > I think it might be safer to move this to setUp and tearDown so it's done per test case rather than per test case class.

Sure, `setUp` and `tearDown` of `AiidaTestCase` already did the new event loop stuff. The exist of the code here is because of the `cls.ids = cls.create_structure_bands()` which is a `ProcessFunction`,  it requires a not closed event loop.  The `setUpClass` is running before the `setUp`, so I have to manually create a event loop. The other option would be remove the ProcessFunction to somewhere else, I tried but failed.

File: aiida/manage/tests/pytest_fixtures.py
Comment (sphuber) [reply aiida/manage/tests/pytest_fixtures.py]: To be honest, I am not sure if it is necessary, but in the scope of testing, it seems good to clean up after oneself. So I would keep it for now

File: setup.json
Comment (sphuber) [inline setup.json:90]:
```suggestion
            "jupyter~=1.0",
            "notebook~=6.0"
```
We are maintaining the policy that if we support a major version of >=1 of a dependency, we always specify `~=1.*` which says basically any minor version of that major series goes, unless we have a very special reason that a certain minor version cannot be used or trusted.

File: aiida/engine/transports.py
Comment (sphuber) [inline aiida/engine/transports.py:102]: Just a quick check: this really needs to be yielded and not awaited?

**Comment by sphuber** [comment]
OK, I think we are almost there @unkcpz and @muhrin . I am just writing down an overview of open questions and things that remain to be checked. Once these are addressed, I will try and run a large production run with the new stack and see how things run.

1. Check correctness of manually cancelling all event loop tasks when closing daemon runner in `aiida/engine/daemon/runner.py `
2. Verify that running a process function no longer needs to create a new runner but just gets the global one: `aiida/engine/processes/functions.py`
3. Find a way for the communicator to always return a similar future, regardless of its type, such that manual unwrapping, as done currently in `aiida/cmdline/commands/cmd_process.py`, is no longer necessary.

@muhrin Have you had a chance to take a look at these three remaining questions?

File: aiida/engine/transports.py
Comment (unkcpz) [reply aiida/engine/transports.py:102]: I think yes, this is the `yield` of contextmanager.

File: setup.json
Comment (unkcpz) [reply setup.json]: Thanks! Changed.

File: aiida/engine/processes/functions.py
Comment (sphuber) [reply aiida/engine/processes/functions.py:116]: > I'm not sure which 'blocking issues' you're referring to @unkcpz but indeed we now only have the one event loop so never need to create more than one. Does `get_runner` work if there is no event loop active at all or would this code be unreachable if that were the case?

`Manager.get_runner` will create a new runner if it doesn't have one yet. It won't pass in a loop to the `Runner` constructor, so that one will get whatever loop the constructor logic defines.

**Comment by sphuber** [comment]
I just created a clean env with this branch and the asyncio compatible plumpy and kiwipy versions and ran a bunch of workchains to test the new code. One of the workchains (the first one that was launched, not sure if by accident) excepted:
```
state        Excepted <Traceback (most recent call last):
               File "/home/sph/.virtualenvs/aiida_pies/lib/python3.7/site-packages/plumpy/processes.py", line 474, in _process_scope
                 yield
               File "/home/sph/.virtualenvs/aiida_pies/lib/python3.7/site-packages/plumpy/processes.py", line 496, in _run_task
                 result = await coro(*args, **kwargs)
             GeneratorExit

             During handling of the above exception, another exception occurred:

             AssertionError: Somehow, the process at the top of the stack is not me, but another process! (<CifCleanWorkChain> (ProcessState.WAITING) != <CifCleanWorkChain> (ProcessState.RUNNING))
             >
```
This clearly comes from the process stack that checks for consistency. If you have ideas what could have caused this, let me know and I can try to dig up more debug information or try to create a MWE.

**Comment by sphuber** [comment]
Tests are failing because it is failing to install the dependencies, especially, it claims there is a dependency conflict between `kiwipy` and, hang on to your hats, `kiwipy`!
```
ERROR: Cannot install kiwipy==0.6.1 and kiwipy[rmq] 0.6.0 because these package versions have conflicting dependencies.

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
The conflict is caused by:
    The user requested kiwipy==0.6.1
    kiwipy[rmq] 0.6.0 depends on kiwipy 0.6.0 (from https://files.pythonhosted.org/packages/fb/5b/9ab8910fcb7b861ae63f0c571bf2be790734a12487c3e5358d6610eef117/kiwipy-0.6.0-py2.py3-none-any.whl#sha256=2e68ba8a74bffb9319c48aa4a33ab78ba5f4c460a1ce7b39092ea2183eb857e5 (from https://pypi.org/simple/kiwipy/) (requires-python:>=3.5))
```
It's you!

Any of the packaging experts out there know what the hell is going on. @chrisjsewell @greschd @csadorf ?

**Comment by chrisjsewell** [comment]
What happens if you remove the use of the new resolver `pip install --use-feature=2020-resolver` at https://github.com/unkcpz/aiida_core/blob/c8a1d89bfd5396e92da0e07e09d2487dbc0794ae/.github/workflows/ci.yml#L123?

**Comment by chrisjsewell** [comment]
Well it installs now 😬  (I guess the pytest failure is a whole other issue).
Interesting that the conda resolver fails in the same way as the new pip resolver; I wonder if the implementation "borrows" from conda.

Hey @pradyunsg sorry for the spam, but I wonder if I can quickly pick your brain quickly on this install issue: https://github.com/aiidateam/aiida-core/pull/4317#issuecomment-687341421.
Does this look like a misunderstanding on our part, or should we raise an issue on pip?

**Comment by pradyunsg** [comment]
>  Does this look like a misunderstanding on our part, or should we raise an issue on pip?

That looks like a bug -- one that has been fixed in pip's master branch. Could you try running off of pip's master branch and check if this bug is reproducible?

**Comment by sphuber** [comment]
>  I wonder if the implementation "borrows" from conda.

When I look at the logs, I see that conda reports a whole shit ton of dependency conflicts though and not just with `kiwipy/plumpy`? I never use conda so I have no clue, but is that normal or to be expected?

**Comment by sphuber** [comment]
> That looks like a bug -- one that has been fixed in pip's master branch. Could you try running off of pip's master branch and check if this bug is reproducible?

Thanks for the comment. We currently install `pip` as `pip install --upgrade pip`. Should I just replace that with something like `pip install --upgrade git+https://github.com/pypa/pip@master` ? Will that properly override the currently installed version?

**Comment by pradyunsg** [comment]
> pip install --upgrade git+https://github.com/pypa/pip@master

`pip install https://github.com/pypa/pip/archive/master.zip` would almost certainly perform better in more ways than one. The 2020-resolver is still under development, so it's still evolving slowly. :)

**Comment by sphuber** [comment]
> `pip install https://github.com/pypa/pip/archive/master.zip` would almost certainly perform better in more ways than one. The 2020-resolver is still under development, so it's still evolving and getting better at being a reasonable resolver. :)

Thanks a lot @pradyunsg and not to worry, fully understand that such a big project takes time to iron out some bugs. Happy to provide bug reports if possible. I tried installing of master but that does not seem to have fixed the problem: https://github.com/aiidateam/aiida-core/pull/4317/checks?check_run_id=1075016847  The install on Python 3.8 fails with the same dependency conflict error. The builds on Python 3.5 are still running and I suspect they will actually time out.

**Comment by chrisjsewell** [comment]
> I never use conda so I have no clue, but is that normal or to be expected?

Yeh that does not look pretty 😬 
It feels like maybe conda, and perhaps the new pip resolver, is not loving the `~=` type version pinning.
You could try replacing them in environment.yml with `>=x.x.x,<y.y.y` type versions?

**Comment by sphuber** [comment]
> It feels like maybe conda, and perhaps the new pip resolver, is not loving the `~=` type version pinning.
> You could try replacing them in environment.yml with `>=x.x.x,<y.y.y` type versions?

Sure I could try that, it's just that we have been using this before without issue. This PR is not the first time we are adding the compatibility operator. So why would that be a problem now? Although, this may be the first versions of `kiwipy` and `plumpy` where we _also_ use that operator. Maybe that is the difference? It struggles when dependencies also use the compat op?

**Comment by greschd** [comment]
@sphuber what happens if you remove the `[rmq]` bit on each part? It kinda looks as if it thinks `kiwipy` and `kiwipy[rmq]` are different packages.

**Comment by pradyunsg** [comment]
Huh! I can't really tell more without taking a closer look, which I don't want to on a weekend. 😅

Consider filing an issue on pip's issue tracker with instructions for reproducing this conflict failure. :)

**Comment by muhrin** [comment]
@unkcpz I've ran the `test_daemon.py` script a few times successfully now using your `plumpy` change.  I _think_ we may be good to go.  Let's see if @sphuber has anything he wants to add but I think it'd be great to get this puppy, finally merged :) Of course, we'll need to go through releasing the new plumpy first, but that's pretty easy.

**Comment by sphuber** [comment]
I will try and run my production test with the new setup a.s.a.p. but will most likely be not until early next week. Keep you posted

**Comment by sphuber** [comment]
@ltalirz the conda install tests are failing because of dependency conflicts. I vaguely remember this happening before but not sure what was done to fix it. Do you have any idea what the cause is and how this is addressed?

**Comment by ltalirz** [comment]
@sphuber I agree this is a bit weird, since supposedly all the new dependencies are available on the conda-forge channel:

aio-pika~=6.6, circus~=0.17.1, kiwipy[rmq]~=0.6.1, plumpy~=0.17.0

I don't know exactly what the problem is; @astamminger has in the past often figured it out (he needs to let us know his tricks at some point ;-) )

My only suspicion at the moment is that this has to do with the fact that there is no PyPy build for aoi-pika https://github.com/conda-forge/aio-pika-feedstock/pulls

However, looking at their meta.yaml https://github.com/conda-forge/aio-pika-feedstock/blob/master/recipe/meta.yaml 
it can actually be converted to a noarch package, so I'll make a PR there

**Comment by sphuber** [comment]
@ltalirz please don't merge develop into this branch. This will have to be rebase-merged and so we need to rebase to update.

**Comment by ltalirz** [comment]
Oops, sorry!

**Comment by ltalirz** [comment]
That doesn't seem to have been the issue :-(

I've looked through the conflict resolution report and it's really not obvious to me where the problem lies - @astamminger any ideas?

**Comment by sphuber** [comment]
@csadorf this is the PR where the conda test install fails.

**Comment by csadorf** [comment]
> @csadorf this is the PR where the conda test install fails.

@sphuber Conda is complaining about a dependency conflict. Unfortunately, [the nonsensical way that conflicts are reported](https://github.com/conda/conda/issues/9707) makes it very difficult to impossible to understand where the conflict originates. I just spent over an hour trying to figure it out and either I just really don't know how to interpret the output or the solver is simply faulty, because I cannot extract a single "real conflict" from the error message.

Do you maybe have some intuition for what might be causing a conflict?

**Comment by ltalirz** [comment]
@sphuber Since from my perspective this is really an issue with the conda infrastructure itself, I'm also ok with allowing the conda-check to fail for some time until we have figured this out.
I would not postpone merging this important PR because of this.

**Comment by unkcpz** [comment]
> Do you maybe have some intuition for what might be causing a conflict?

@csadorf The new package introduced here is `aio-pika`. Maybe it is the last message(below) which indicates where the issues happened?
```
Package pamqp conflicts for:
aio-pika~=6.6 -> aiormq[version='>=3.2.3,<4'] -> pamqp==2.3.0
pamqp~=2.0
```

**Comment by sphuber** [comment]
> Do you maybe have some intuition for what might be causing a conflict?

Not really, I also came to the same conclusion as yourself that the report of conflict is anything but clear and could not figure out the actual cause. I never use conda myself and so have very little experience with it. What @unkcpz mentioned might be the real hint though: we indeed got new dependencies in this PR [`aio-pika`, `aiormq` and `pamqp`] from which only the last I added as an explicit dependency to `aiida-core` because that is the only library that we call expliciltly. However, `aiormq` also directly depends on this and so the conflict may come from there.

**Comment by sphuber** [comment]
> I would not postpone merging this important PR because of this.

Agreed. I was going to merge it, however, since this affects a lot how connections and interactions with RabbitMQ are performed (we use completely different libraries with this PR) I wanted to add some automated tests. I manually found a bug that I fixed with the last commit on this PR, but it completely went unnoticed by our CI. Ideally then, I would first get #4579 merged so we actually test this failure. Ideally this would also include tests for SSL, for which I had to debug a lot in `aiida-core==v1.4.*` but I am not sure this is feasible. Anyway, I will merge this next week when I get back from holidays

**Comment by ltalirz** [comment]
> > Do you maybe have some intuition for what might be causing a conflict?
> 
> @csadorf The new package introduced here is `aio-pika`. Maybe it is the last message(below) which indicates where the issues happened?
> 
> ```
> Package pamqp conflicts for:
> aio-pika~=6.6 -> aiormq[version='>=3.2.3,<4'] -> pamqp==2.3.0
> pamqp~=2.0
> ```

Thanks @unkcpz for checking.

In principle, `==2.3.0` is compatible with `~=2.0`, but I agree that looking into `aio-pika` and its dependencies may lead to the solution.

For pamqp, both 2.3.0 and 3.0.1 are available and they are `noarch`, so there should be no compatibility issues
https://anaconda.org/conda-forge/pamqp/files

**Comment by csadorf** [comment]
I've removed a few dependencies, including `aio-pika` and `pamqp`, with no luck. @unkcpz How was this `environment.yml` file created? Have you ever successfully created a conda environment without conflicts or did you just update the file?

**Comment by ltalirz** [comment]
@csadorf the `environment.yml` is created automatically from the specified dependencies

**Comment by csadorf** [comment]
> @csadorf the `environment.yml` is created automatically from the specified dependencies

So, in other words, we are not sure that a conflict-free *conda* environment exists, right?

**Comment by unkcpz** [comment]
@csadorf @ltalirz I think I find which two package conflict but have no idea what's going on. 
Either I move `python~=3.7` or `plumpy~=0.17.0` there will be no complaint of conflicting.

**Comment by sphuber** [comment]
Does anyone know when it says `python~=3.7` when we also still support 3.6? Is that the problem?

**Comment by csadorf** [comment]
> Does anyone know when it says `python~=3.7` when we also still support 3.6? Is that the problem?

The `environment.yml` file is a utility file that users should not need to worry about and that has no influence into which environment aiida-core can be installed. We use it to test whether AiiDA is install-able into a valid conda environment, but a user would just say `conda install aiida-core` or create an environment as described in our [docs](https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/install_conda.html). When you install a package with conda, the Python version is usually implicitly determined from your base environment, but you can explicitly select a specific Python version by adding it as an explicit dependency, i.e., `conda create -n aiida-py38 python=3.8 aiida-core`.

So, I don't think that we *need* to have the Python version specified in `environment.yml`, but it is concerning that removing it apparently resolves the problem as it hints as to an issue that is specific to Python 3.7.

@unkcpz Which Python version is chosen when you remove the explicit dependency? Also, could you please share your environment with us after you successfully created it as described [here](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#sharing-an-environment)?

**Comment by ltalirz** [comment]
Thanks a lot @unkcpz for figuring this out!

For reference, here is the commit where I introduced the python version in the `environment.yml` https://github.com/aiidateam/aiida-core/commit/12f96415e48a43fa43f6b2493101cb5e8b27108a
As the commit message says, we figured out that when we were specifying the python version of the environment via the command line, it was not respected.
I just checked and this is still the case with conda 4.9.1
```
$ conda  env create -f environment.yml -n test-conda python=3.5
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.9.1
  latest version: 4.9.2

Please update conda by running

    $ conda update -n base -c defaults conda



Downloading and Extracting Packages
tzdata-2020d         | 110 KB    | ################################################################################################################################################################### | 100%
certifi-2020.11.8    | 148 KB    | ################################################################################################################################################################### | 100%
wheel-0.35.1         | 38 KB     | ################################################################################################################################################################### | 100%
setuptools-50.3.1    | 726 KB    | ################################################################################################################################################################### | 100%
pip-20.2.4           | 1.7 MB    | ################################################################################################################################################################### | 100%
python-3.9.0         | 9.9 MB    | ################################################################################################################################################################### | 100%
```
I.e. specifying the python version inside the `environment.yml` was a way for us to make the build of the environment on CI reproducible.

python 3.7 was probably chosen because it was the latest python release that we tested on, i.e. the version number can be changed (it could even be changed to python 3.6 if that helps).

The commit message also mentions that at that time there had been an issue with plumpy, which had not yet been marked as compatible with python 3.7 - but that issue had been fixed before this commit (perhaps coincidence).

**Comment by unkcpz** [comment]
> @unkcpz Which Python version is chosen when you remove the explicit dependency? Also, could you please share your environment with us after you successfully created it as described here?

Here as @ltalirz mentioned, I run `conda  env create -f environment.yml -n test-conda python=3.5` to create the envs locally. All python versions specified are good. 
Changing `python~=3.7` to `python~=3.6` in the file really solves the problem. So we adopt this as workaround?

**Comment by unkcpz** [comment]
The problem may be caused by the package `aiocontextvars` which is used to support to the PEP-567 that is 100% contextvars provided by std library, you can check its dependencies here at https://github.com/fantix/aiocontextvars/blob/master/setup.py. 

This dependency can be removed once we no longer support python3.6.

**Comment by csadorf** [comment]
@unkcpz Thx for identifying the issue! It would still be great if you could share your `environment.yml` for a successful installation. Thx!

**Comment by unkcpz** [comment]
@csadorf Sure, but I don't think it helps, in my local machine, I only use conda to create the env and install the packages by `pip`

<details><summary>environment.yml</summary>
<p>

```
name: /data/CONDA_ENV/aiida-env-testing
channels:
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - backcall=0.2.0=py_0
  - binutils_impl_linux-64=2.33.1=he6710b0_7
  - binutils_linux-64=2.33.1=h9595d00_15
  - ca-certificates=2020.10.14=0
  - certifi=2020.11.8=py38h06a4308_0
  - decorator=4.4.2=py_0
  - gcc_impl_linux-64=7.3.0=habb00fd_1
  - gcc_linux-64=7.3.0=h553295d_15
  - gxx_impl_linux-64=7.3.0=hdf63c60_1
  - gxx_linux-64=7.3.0=h553295d_15
  - ipython=7.19.0=py38hb070fc8_0
  - ipython_genutils=0.2.0=py38_0
  - jedi=0.17.2=py38_0
  - ld_impl_linux-64=2.33.1=h53a641e_7
  - libedit=3.1.20191231=h14c3975_1
  - libffi=3.3=he6710b0_2
  - libgcc-ng=9.1.0=hdf63c60_0
  - libstdcxx-ng=9.1.0=hdf63c60_0
  - ncurses=6.2=he6710b0_1
  - openssl=1.1.1h=h7b6447c_0
  - parso=0.7.0=py_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=py38_1000
  - pip=20.2.4=py38h06a4308_0
  - prompt-toolkit=3.0.8=py_0
  - ptyprocess=0.6.0=pyhd3eb1b0_2
  - pygments=2.7.2=pyhd3eb1b0_0
  - python=3.8.5=h7579374_1
  - readline=8.0=h7b6447c_0
  - setuptools=50.3.1=py38h06a4308_1
  - sqlite=3.33.0=h62c20be_0
  - tk=8.6.10=hbc83047_0
  - traitlets=5.0.5=py_0
  - wcwidth=0.2.5=py_0
  - wheel=0.35.1=pyhd3eb1b0_0
  - xz=5.2.5=h7b6447c_0
  - zlib=1.2.11=h7b6447c_3
  - pip:
    - aiida-core==1.5.0
    - aio-pika==6.7.1
    - aiocontextvars==0.2.2
    - aiormq==3.3.1
    - aldjemy==0.9.1
    - alembic==1.4.3
    - archive-path==0.2.1
    - async-generator==1.10
    - bcrypt==3.2.0
    - cffi==1.14.3
    - circus==0.17.1
    - click==7.1.2
    - click-completion==0.5.2
    - click-config-file==0.6.0
    - click-spinner==0.1.10
    - configobj==5.0.6
    - cryptography==3.2.1
    - deprecation==2.1.0
    - django==2.2.17
    - ete3==3.1.2
    - frozendict==1.2
    - idna==2.10
    - jinja2==2.11.2
    - kiwipy==0.6.1
    - mako==1.1.3
    - markupsafe==1.1.1
    - multidict==5.0.2
    - nest-asyncio==1.4.3
    - numpy==1.19.4
    - packaging==20.4
    - pamqp==2.3.0
    - paramiko==2.7.2
    - pgsu==0.1.0
    - plumpy==0.17.0
    - psutil==5.7.3
    - psycopg2-binary==2.8.6
    - pycparser==2.20
    - pynacl==1.4.0
    - pyparsing==2.4.7
    - python-dateutil==2.8.1
    - python-editor==1.0.4
    - python-graphviz==0.15
    - pytray==0.2.2
    - pytz==2019.3
    - pyyaml==5.1.2
    - pyzmq==20.0.0
    - reentry==1.3.1
    - shellingham==1.3.2
    - shortuuid==1.0.1
    - simplejson==3.17.2
    - six==1.15.0
    - sqlalchemy==1.3.20
    - sqlalchemy-utils==0.34.2
    - sqlparse==0.4.1
    - tabulate==0.8.7
    - tornado==6.1
    - tqdm==4.53.0
    - tzlocal==2.1
    - upf-to-json==0.9.2
    - wrapt==1.11.2
    - yarl==1.6.3
prefix: /data/CONDA_ENV/aiida-env-testing
```

</p>
</details>

**Comment by csadorf** [comment]
> The problem may be caused by the package `aiocontextvars` which is used to support to the PEP-567 that is 100% contextvars provided by `python>=3.7` std library, you can check its dependencies here at https://github.com/fantix/aiocontextvars/blob/master/setup.py.
> 
> This dependency can be removed once we no longer support python3.6.

@unkcpz That is 100% the problem. [This conda package explicitly requires `python<3.7`](https://github.com/conda-forge/aiocontextvars-feedstock/blob/dd7a370ee6f282f53659bdee2e2b2d04a04148e1/recipe/meta.yaml#L17-L24). Which package depends on it?

**Comment by ltalirz** [comment]
@csadorf Turns out it's plumpy https://github.com/conda-forge/plumpy-feedstock/blob/fe935aa06b56595ef3d5b2eaccafbe85e52da892/recipe/meta.yaml#L27

So, plumpy should depend on this package only for python < 3.7 (which will force us to remove the `noarch` from plumpy)

**Comment by unkcpz** [comment]
>  That is 100% the problem. This conda package explicitly requires python<3.7. Which package depends on it?

Thanks @csadorf. `aiocontextvars` is required by `plumpy` to support `contextvar` which is new in python3.7. In order to use this in python3.6 the `aiocontextvars` is needed. For environment py>36 the  the native `contextvars` is called by this package so it support python>3.6 in principle.

**Comment by sphuber** [comment]
> So, plumpy should depend on this package only for python < 3.7 (which will force us to remove the `noarch` from plumpy)
> Can someone please take care of this?

@ltalirz does this only need to happen for the feedstock, or will I have to release a `0.17.1` on PyPI as well where the requirements are updated?

**Comment by chrisjsewell** [comment]
> does this only need to happen for the feedstock, or will I have to release a 0.17.1 on PyPI as well where the requirements are updated?

I think we will need to release as well. While we are at it: gentle prod to review https://github.com/aiidateam/plumpy/pull/180, then I'll be happy to handle these releases

**Comment by csadorf** [comment]
> > does this only need to happen for the feedstock, or will I have to release a 0.17.1 on PyPI as well where the requirements are updated?
> 
> I think we will need to release as well.

I would think just for the feedstock?

**Comment by sphuber** [comment]
> I think we will need to release as well. While we are at it: gentle prod to review [aiidateam/plumpy#180](https://github.com/aiidateam/plumpy/pull/180), then I'll be happy to handle these releases wink

I will hold off with that until this is merged and stable. The mypy PR contains a lot of changes some and may have an effect on this PR here which may not be tested.

**Comment by unkcpz** [comment]
Thanks @chrisjsewell, I'll get time tomorrow to take care of https://github.com/aiidateam/plumpy/pull/180 . But I agree with Sebastiaan we wait a minute after this PR.

**Comment by chrisjsewell** [comment]
> The mypy PR contains a lot of changes some and may have an effect on this PR here which may not be tested.

Fair 👍
(although in principle it should only break this PR if the PR is already broken, i.e. has faulty types)

**Review by sphuber — APPROVED** [review]

**Pull request merged**
